lib/caching-precompiler.js

```diff
@@ -0,0 +1,95 @@
+var cachingTransform = require('caching-transform');
+var fs = require('fs');
+var path = require('path');
+var md5Hex = require('md5-hex');
+var stripBom = require('strip-bom');
+
+module.exports = CachingPrecompiler;
+
+function CachingPrecompiler(cacheDir) {
+	if (!(this instanceof CachingPrecompiler)) {
+		throw new Error('CachingPrecompiler must be called with new');
+	}
+	this.cacheDir = cacheDir;
+	this.filenameToHash = {};
+	this.transform = this._createTransform();
+}
+
+CachingPrecompiler.prototype._factory = function (cacheDir) {
+	// This factory method is only called once per process, and only as needed, to defer loading expensive dependencies.
+	var babel = require('babel-core');
+	var convertSourceMap = require('convert-source-map');
+	var presetStage2 = require('babel-preset-stage-2');
+	var presetES2015 = require('babel-preset-es2015');
+	var transformRuntime = require('babel-plugin-transform-runtime');
+
+	var powerAssert = this._createEspowerPlugin(babel);
+
+	function buildOptions(filename, code) {
+		// Extract existing source maps from the code.
+		var sourceMap = convertSourceMap.fromSource(code) || convertSourceMap.fromMapFileSource(code, path.dirname(filename));
+
+		return {
+			presets: [presetStage2, presetES2015],
+			plugins: [powerAssert, transformRuntime],
+			filename: filename,
+			sourceMaps: true,
+			ast: false,
+			inputSourceMap: sourceMap && sourceMap.toObject()
+		};
+	}
+
+	return function (code, filename, hash) {
+		code = code.toString('utf8');
+		var options = buildOptions(filename, code);
+		var result = babel.transform(code, options);
+		var mapFile = path.join(cacheDir, hash + '.map');
+		fs.writeFileSync(mapFile, JSON.stringify(result.map));
+		return result.code;
+	};
+};
+
+CachingPrecompiler.prototype._createEspowerPlugin = function (babel) {
+	var createEspowerPlugin = require('babel-plugin-espower/create');
+	var enhanceAssert = require('./enhance-assert');
+
+	// initialize power-assert
+	return createEspowerPlugin(babel, {
+		patterns: enhanceAssert.PATTERNS
+	});
+};
+
+CachingPrecompiler.prototype._createTransform = function () {
+	var self = this;
+
+	return cachingTransform({
+		factory: this._factory.bind(this),
+		cacheDir: self.cacheDir,
+		salt: new Buffer(JSON.stringify({
+			'babel-plugin-espower': require('babel-plugin-espower/package.json').version,
+			'ava': require('../package.json').version,
+			'babel-core': require('babel-core/package.json').version
```

Comment:
Those are just presets, and don't reflect the actual versions of the underlying plugins (they use caret ranges). While it won't break anything to salt with preset versions, it may be providing a false sense of security.

@thejameskyle - Any thoughts on this? Last time I browsed the babel source, it seemed you were only including the version of `babel-core` in the salt for the cache key.